src/js/model/index.js

```diff
@@ -331,26 +345,78 @@ function updateSelectedMarkInVega() {
  * naively set all signals on each store update to ensure store and Vega
  * stay in sync. Altering this to be smarter is one area to investigate should
  * any performance issues crop up later on, but signal writes are pretty fast.
+ *
  * @returns {void}
  */
-function updateAllSignals() {
-  // Nothing to do here if the view is not ready
-  if (!model.view || typeof model.view.signal !== 'function') {
-    return;
-  }
-  var signals = getIn(store.getState(), 'signals');
-  signals.forEach(function(value, name) {
+function updateAllSignals(state) {
+  getIn(state, 'signals').forEach(function(value, name) {
     // Skip any signal from the defaults
     if (sg.isDefault(name)) {
       return;
     }
-    // Persist any signals from the store to the view
-    model.view.signal(name, value.init);
+    // Persist any signals from the store to the view: wrap this in a try/catch
+    // to handle situations where the update fires while we are registering
+    // signals for new marks before a reparse has actually injected those marks
+    // into the vega view itself.
+    try {
+      model.view.signal(name, value.init);
+    } catch (e) { /* This is ok */ }
+  });
+}
+
+function initiateReparse() {
+  model.parse().then(function() {
+    store.dispatch(parseComplete());
   });
 }
+var debouncedInitiateReparse = debounce(initiateReparse, 16);
+
+function recreateVegaIfNecessary(state) {
+  var shouldReparse = getIn(state, 'viewState.reparseModel');
+
+  if (shouldReparse) {
+    if (model.view) {
+      // Clear out the outdated vega spec: iterate through all registered
+      // signal streams and remove their event listeners
+      model.view.destroy();
+      model.view = null;
+    }
+    store.dispatch(parseStart());
+    // Don't start a reparse more often than 60 times a second
+    debouncedInitiateReparse();
+    // initiateReparse();
```

Comment:
This is the component of this about which I am least satisfied: if we do not debounce this, then the pipelines sometimes fail to load. Why, is anybody's guess. Hopefully moving that sort of action into the store will fix this naturally later on.